unix-domain-socket/src/main/scala/akka/stream/alpakka/unixdomainsocket/scaladsl/UnixDomainSocket.scala

```diff
@@ -0,0 +1,515 @@
+/*
+ * Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.unixdomainsocket.scaladsl
+
+import java.io.File
+import java.nio.ByteBuffer
+import java.nio.channels.{SelectionKey, Selector}
+
+import akka.{Done, NotUsed}
+import akka.actor.{
+  ActorSystem,
+  Cancellable,
+  CoordinatedShutdown,
+  ExtendedActorSystem,
+  Extension,
+  ExtensionId,
+  ExtensionIdProvider
+}
+import akka.stream._
+import akka.stream.scaladsl.{BroadcastHub, Flow, Keep, Sink, Source, SourceQueueWithComplete}
+import akka.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}
+import akka.util.ByteString
+import jnr.enxio.channels.NativeSelectorProvider
+import jnr.unixsocket.{UnixServerSocketChannel, UnixSocketAddress, UnixSocketChannel}
+
+import scala.concurrent.duration.{Duration, FiniteDuration}
+import scala.concurrent.{ExecutionContext, Future, Promise}
+import scala.util.{Failure, Success, Try}
+import scala.util.control.NonFatal
+
+object UnixDomainSocket extends ExtensionId[UnixDomainSocket] with ExtensionIdProvider {
+
+  def apply()(implicit system: ActorSystem): UnixDomainSocket = super.apply(system)
+
+  override def createExtension(system: ExtendedActorSystem) =
+    new UnixDomainSocket(system)
+
+  override def lookup(): ExtensionId[_ <: Extension] =
+    UnixDomainSocket
+
+  /**
+   * * Represents a successful server binding.
+   */
+  final case class ServerBinding(localAddress: UnixSocketAddress)(private val unbindAction: () => Future[Unit]) {
+    def unbind(): Future[Unit] = unbindAction()
+  }
+
+  /**
+   * Represents an accepted incoming connection.
+   */
+  final case class IncomingConnection(localAddress: UnixSocketAddress,
+                                      remoteAddress: UnixSocketAddress,
+                                      flow: Flow[ByteString, ByteString, NotUsed]) {
+
+    /**
+     * Handles the connection using the given flow, which is materialized exactly once and the respective
+     * materialized instance is returned.
+     *
+     * Convenience shortcut for: `flow.join(handler).run()`.
+     */
+    def handleWith[Mat](handler: Flow[ByteString, ByteString, Mat])(implicit materializer: Materializer): Mat =
+      flow.joinMat(handler)(Keep.right).run()
+  }
+
+  /**
+   * Represents a prospective outgoing TCP connection.
+   */
+  final case class OutgoingConnection(remoteAddress: UnixSocketAddress, localAddress: UnixSocketAddress)
+
+  private val ReceiveBufferSize = 8192
```

Comment:
Magic number? Should that be configurable? If not, a comment about the value woule be nice.

Reply:
It should ultimately be configurable. I've actually now increased it to 64k which is fine given the lack of a network - it could probably even go to about 200k. My Linux installation states 208k is the current max buffer size for sending.

Reply:
I've also documented my rationale as a comment.